sophiread/FastSophiread/src/tpx3_fast.cpp

```diff
@@ -60,6 +77,55 @@ std::vector<TPX3> findTPX3H(ForwardIter begin, ForwardIter end) {
   return batches;
 }
 
+/**
+ * @brief Templated function to locate all TPX3H (chip dataset) in the raw data.
+ *
+ * @tparam ForwardIter
+ * @param[in] begin
+ * @param[in] end
+ * @param[out] consumed
+ * @return std::vector<TPX3>
+ * @note will limit the batch size to 100000, will update the number of elements consumed
+ */
+template <typename ForwardIter>
+std::vector<TPX3> findTPX3H(ForwardIter begin, ForwardIter end, std::size_t &consumed) {
+  std::vector<TPX3> batches;
+  auto len = std::distance(begin, end) / 64;
+#ifdef MAX_BATCH_LEN
+  auto _max_batch_len = _get_max_batch_len();
+  if ( (long unsigned int)len > _max_batch_len ) {
+    len = _max_batch_len;
+  }
+#endif  // MAX_BATCH_LEN
+  batches.reserve(len);
+  consumed = 0;
+
+  // local variables
+  int chip_layout_type = 0;
+  int data_packet_size = 0;
+  int data_packet_num = 0;
+
+  // find all batches
+  for (auto iter = begin; std::distance(iter, end) >= 8; std::advance(iter, 8), consumed += 8) {
+    const char *char_array = &(*iter);
+#ifdef MAX_BATCH_LEN
+    if (batches.size() >= _max_batch_len) {
+      break;
```

Comment:
can we throw a runtime warning to indicate insufficient batch size?

Reply:
You want to emit an `spdlog.warning()`?  An exception would not be appropriate here, since batching is *supposed* to limit the size of your batch, thus causing you to iterate.